File: leetcode/1049.cpp
class Solution {
public:
    int lastStoneWeightII(vector<int>& stones) {
        int sum = accumulate(stones.begin(), stones.end(), 0);
        int t = sum / 2;
        int dp[t + 1];
        memset(dp, 0, sizeof dp);
        for (auto stone : stones) {
            for (int i = t; i >= stone; --i) {
                dp[i] = max(dp[i], dp[i - stone] + stone);
            }
        }
        return sum - 2 * dp[t];
    }
};

File: leetcode/1049.go
package leetcode

func lastStoneWeightII(stones []int) int {
	sum := 0
	for _, v := range stones {
		sum += v
	}
	m := sum / 2
	dp := make([]int, m+1)
	for _, weight := range stones {
		for j := m; j >= weight; j-- {
			dp[j] = max(dp[j], dp[j-weight]+weight)
		}
	}
	return sum - 2*dp[m]
}
func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
